ndk/src/font.rs

```diff
@@ -0,0 +1,401 @@
+    pub fn new() -> Self {
+        NonNull::new(unsafe { ffi::AFontMatcher_create() })
+            .map(|p| unsafe { FontMatcher::from_ptr(p) })
+            .expect("AFontMatcher_create returned NULL")
+    }
+
+    /// Set family variant to matcher.
+    ///
+    /// If this function is not called, the matcher performs with [`FamilyVariant::Default`].
+    pub fn set_family_variant(&mut self, family_variant: FamilyVariant) {
+        unsafe { ffi::AFontMatcher_setFamilyVariant(self.ptr.as_ptr(), family_variant as u32) }
+    }
+
+    /// Set font locales to matcher.
+    ///
+    /// If this function is not called, the matcher performs with empty locale list.
+    ///
+    /// # Arguments
+    /// * `language_tags`: a null character terminated comma separated IETF BCP47 compliant language tags.
+    pub fn set_locales(&mut self, language_tags: &CStr) {
+        unsafe { ffi::AFontMatcher_setLocales(self.ptr.as_ptr(), language_tags.as_ptr()) }
+    }
+
+    /// Set font style to matcher.
+    ///
+    /// If this function is not called, the matcher performs with [`FontWeight::NORMAL`] with non-italic style.
+    pub fn set_style(&mut self, weight: FontWeight, italic: bool) {
+        unsafe { ffi::AFontMatcher_setStyle(self.ptr.as_ptr(), weight.value(), italic) }
+    }
+}
+
+#[cfg(feature = "api-level-29")]
+impl Drop for FontMatcher {
+    fn drop(&mut self) {
+        unsafe { ffi::AFontMatcher_destroy(self.ptr.as_ptr()) }
+    }
+}
+
+/// A native [`ASystemFontIterator *`]
+///
+/// [`ASystemFontIterator *`]: https://developer.android.com/ndk/reference/group/font#asystemfontiterator_open
+#[cfg(feature = "api-level-29")]
+#[derive(Debug)]
+pub struct SystemFontIterator {
+    ptr: NonNull<ffi::ASystemFontIterator>,
+}
+
+#[cfg(feature = "api-level-29")]
+impl SystemFontIterator {
+    /// Create an `SystemFontIterator` from a pointer
+    ///
+    /// # Safety
+    /// By calling this function, you assert that the pointer is a valid pointer to a native
+    /// `ASystemFontIterator`.
+    pub unsafe fn from_ptr(ptr: NonNull<ffi::ASystemFontIterator>) -> Self {
+        Self { ptr }
+    }
+
+    /// Returns the pointer to the native `ASystemFontIterator`.
+    pub fn ptr(&self) -> NonNull<ffi::ASystemFontIterator> {
+        self.ptr
+    }
+
+    /// Create a system font iterator.
+    pub fn new() -> Option<Self> {
+        NonNull::new(unsafe { ffi::ASystemFontIterator_open() })
+            .map(|p| unsafe { SystemFontIterator::from_ptr(p) })
+    }
+}
+
+#[cfg(feature = "api-level-29")]
+impl Iterator for SystemFontIterator {
+    type Item = Font;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        NonNull::new(unsafe { ffi::ASystemFontIterator_next(self.ptr.as_ptr()) })
+            .map(|p| unsafe { Font::from_ptr(p) })
+    }
+}
+
+#[cfg(feature = "api-level-29")]
+impl Drop for SystemFontIterator {
+    fn drop(&mut self) {
+        unsafe { ffi::ASystemFontIterator_close(self.ptr.as_ptr()) }
```

Comment:
This is documented as "freeing any related resources". Does that include the fonts that are returned, or should the caller clean those up manually? (I assume the latter, since this seems to be the only way to get hold of an `AFont` object, and that is what the return docs for `ASystemFontIterator_next()` somewhat clarify)

Reply:
I also assumed the latter because the current code (https://cs.android.com/android/platform/superproject/+/master:frameworks/base/native/android/system_fonts.cpp;l=373) returns an owning pointer.